src/pool/srv_pool_scrub.c

```diff
@@ -4,324 +4,771 @@
+				      entry->ie_minor_epc)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			ctx->sc_epoch = 0;
+			ctx->sc_minor_epoch = 0;
+		} else {
+			C_TRACE("Scrubbing akey: "DF_KEY", type: %s, rec size: "
+					DF_U64", extent: "DF_RECX"\n",
+				DP_KEY(&param->ip_akey),
+				(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
+				entry->ie_rsize,
+				DP_RECX(entry->ie_orig_recx)
+			);
+
+			ctx->sc_iod.iod_size = entry->ie_rsize;
+			ctx->sc_iod.iod_nr = 1;
+			ctx->sc_iod.iod_recxs = &entry->ie_recx;
+			ctx->sc_iod.iod_type = type == VOS_ITER_RECX ?
+					       DAOS_IOD_ARRAY :
+					       DAOS_IOD_SINGLE;
+			sc_obj_val_setup(ctx, entry, type, param, ih);
+
+			rc = sc_verify_obj_value(ctx);
+			*acts |= VOS_ITER_CB_YIELD;
+
+			if (rc != 0) {
+				D_ERROR("Error Verifying:"DF_RC"\n", DP_RC(rc));
+				return rc;
+			}
+			ds_scrub_sched_control(ctx);
+		}
+		break;
+	}
 
-	if (ctx->msec_between_calcs == 0) {
-		C_TRACE("Yield after data scrub\n");
-		dss_ult_yield(ctx->req);
-	} else {
-		C_TRACE("Sleeping after data scrub for "DF_U64" msec\n",
-			ctx->msec_between_calcs);
-		sched_req_sleep(ctx->req, ctx->msec_between_calcs);
+	case VOS_ITER_DTX:
+	case VOS_ITER_COUUID:
+	case VOS_ITER_NONE:
+		D_ASSERTF(0, "Invalid.");
 	}
 
 	return 0;
 }
 
+static void
+sc_add_cont_metrics(struct scrub_ctx *ctx)
+{
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_last_duration,
+			D_TM_DURATION,
+			"How long the previous scrub took", "ms",
+			DF_CONT_DIR"/"M_LAST_DURATION,
+			DP_UUID((ctx)->sc_pool_uuid),
+			(ctx)->sc_dmi->dmi_tgt_id,
+			DP_UUID(ctx->sc_cont.scs_cont_uuid));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_start,
+			D_TM_TIMESTAMP,
+			"When the current scrubbing started", NULL,
+			DF_CONT_DIR"/"M_STARTED, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_csum_calcs,
+			D_TM_COUNTER, "Number of checksums calculated for "
+				      "current scan",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_COUNTER, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_last_csum_calcs,
+			D_TM_COUNTER, "Number of checksums calculated in last "
+				      "scan", NULL,
+			DF_CONT_DIR"/"M_CSUM_PREV_COUNTER,
+			DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_total_csum_calcs,
+			D_TM_COUNTER, "Total number of checksums calculated",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_TOTAL_COUNTER, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_corruption,
+			D_TM_COUNTER, "Number of silent data corruption "
+				      "detected during current scan",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_CORRUPTION, DP_CONT_DIR(ctx));
+	d_tm_add_metric(&ctx->sc_metrics.scm_cont_metrics.sm_total_corruption,
+			D_TM_COUNTER, "Total number of silent data corruption "
+				      "detected",
+			NULL,
+			DF_CONT_DIR"/"M_CSUM_TOTAL_CORRUPTION,
+			DP_CONT_DIR(ctx));
+}
+
 static int
-scrub_cont(struct scrub_ctx *ctx)
+sc_scrub_cont(struct scrub_ctx *ctx)
 {
 	vos_iter_param_t		param = {0};
```

Comment:
I don't completely follow. Why isn't every record iterated?

Reply:
evtree isn't a sorted tree, to iterate evtree like a sorted tree, we temporarily generate an in-memory auxiliary sorted index for sorted evtree iterator, to use sorted iterator, one needs to specify VOS_IT_RECX_VISIBLE/COVERED (see vos_aggregate()), otherwise, unsorted iterator which just iterates the evtree in physical order will be used.

Imagine the scrub unsorted iterator yield in iterate callback, and the current record (the anchor) is deleted by aggregation, when the iterator resume to re-probe and failed to find the anchor, it won't be able to proceed since it has no idea what's the 'next' item.

I'm wondering how should we deal with this, let the scrubber start over once re-probe failed with nonexistent? Or should we only scrubbing the visible records (only top view, snapshots are visible to app, records being covered are actually not visible to app and it's kind of wasteful to scrub them) in sorted iterator somehow? Not matter which approach is used, seems we have to deal with the race between csum scrubber and aggregation.

Reply:
I see. Would you be okay if I created a defect for this and we moved forward trying to get this PR landed? There are some complexities with partial extents that I will have to deal with when the VISIBLE flag is set. As is it should work, even if extents are missed because of the unordered issue. They should get scrubbed during the next pass.